intel-sgx/insecure-time/src/lib.rs

```diff
@@ -0,0 +1,861 @@
+    unsafe { 
+        let res = __cpuid(leaf);
+        *eax = res.eax;
+        *ebx = res.ebx;
+        *ecx = res.ecx;
+        *edx = res.edx;
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Ticks(AtomicU64);
+
+impl PartialEq<Ticks> for Ticks {
+    fn eq(&self, other: &Ticks) -> bool {
+        let t = self.0.load(Ordering::Relaxed);
+        let o = other.0.load(Ordering::Relaxed);
+        t.eq(&o)
+    }
+}
+
+impl PartialOrd<Ticks> for Ticks {
+    fn partial_cmp(&self, other: &Ticks) -> Option<cmp::Ordering> {
+        let t = self.0.load(Ordering::Relaxed);
+        let o = other.0.load(Ordering::Relaxed);
+        t.partial_cmp(&o)
+    }
+}
+
+impl Add for Ticks {
+    type Output = Ticks;
+
+    fn add(self, other: Self) -> Self::Output {
+        Ticks(AtomicU64::new(self.0.load(Ordering::Relaxed) + other.0.load(Ordering::Relaxed)))
+    }
+}
+
+impl Display for Ticks {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(f, "{} ticks", self.0.load(Ordering::Relaxed))
+    }
+}
+
+impl Ticks {
+    pub fn new(t: u64) -> Self {
+        Ticks(AtomicU64::new(t))
+    }
+
+    pub fn now() -> Self {
+        Ticks(Rdtscp::read().into())
+    }
+
+    pub fn abs_diff(&self, t1: &Ticks) -> Ticks {
+        let t0 = self.0.load(Ordering::Relaxed);
+        let t1 = t1.0.load(Ordering::Relaxed);
+
+        if t1 > t0 {
+            Ticks::new(t1 - t0)
+        } else {
+            Ticks::new(t0 - t1)
+        }
+    }
+
+    fn set(&self, t: Ticks) {
+        let t = t.0.load(Ordering::Relaxed);
+        self.0.store(t, Ordering::Relaxed);
+    }
+
+    pub fn from_duration(duration: Duration, freq: &Freq) -> Self {
+        let freq = freq.as_u64();
+        let ticks_secs = duration.as_secs() * freq;
+        let ticks_nsecs = duration.subsec_nanos() as u64 * freq / NANOS_PER_SEC as u64;
+        Ticks::new(ticks_secs + ticks_nsecs)
+    }
+
+    pub fn as_duration_ex(&self, freq: &Freq) -> Duration {
+        let freq = freq.as_u64();
+        let ticks = self.0.load(Ordering::Relaxed);
+
+        let time_secs = ticks / freq;
+        let time_nsecs = (ticks % freq * NANOS_PER_SEC as u64) / freq;
+        let time_nsecs: u32 = time_nsecs.try_into().expect("must be smaller than 1sec");
+
+        Duration::new(time_secs, time_nsecs)
+    }
+
+    pub fn elapsed(&self, freq: &Freq) -> Duration {
+        let tsc_now = Ticks::now();
+        self.abs_diff(&tsc_now).as_duration_ex(&freq)
+    }
+}
+
+pub struct Rdtscp;
+
+impl Rdtscp {
+    #[inline(never)]
+    pub fn read() -> u64 {
+        let mut aux: u32 = 0;
+        unsafe { __rdtscp(&mut aux) }
+    }
+
+    #[cfg(not(target_env = "sgx"))]
+    pub fn is_supported() -> bool {
+        let mut eax = 0;
+        let mut ebx = 0;
+        let mut ecx = 0;
+        let mut edx = 0;
+
+        cpuid(0x12, &mut eax, &mut ebx, &mut ecx, &mut edx);
+        eax & (0x1 << 1) != 0x0
+
+    }
+}
+
+/// Frequency in ticks per second
```

Comment:
From the type, it is clear that the ticks and seconds are aligned (e.g., there is always an integer amount of ticks per second). Is this always the case? If so, can you add a comment? If not, do we need to handle carry ticks or is this effect negligible?

Reply:
(Some) hardware does report the frequency of the crystal clock. Interestingly, using that frequency directly results in a small (non negligible) clock drift. To account for that, there's a way to "learn" the correct frequency by requesting the correct wallclock time through a usercall, and reading the value of the TSC (see `TscMode::Learn`). There's also an option to ensure that the clock is re-synced within a specific time interval. That way clock drift is kept within limits.
See also the internal "Insecure time without leaving the enclave" page.